src/frame-writer.cpp

```diff
@@ -523,94 +525,216 @@ static enum AVSampleFormat convert_codec_sample_fmt(const AVCodec *codec, std::s
+    inputFrame->channel_layout = AV_CH_LAYOUT_STEREO;
+    inputFrame->channels = 2;
+#endif
+    inputFrame->format = AV_SAMPLE_FMT_FLT;
+    inputFrame->nb_samples = len / (sizeof(float) * 2); // 2 channels
+
+    av_frame_get_buffer(inputFrame, 0);
+    memcpy(inputFrame->data[0], data, len);
+
+    // Create output frame with codec's parameters
+    AVFrame *outputFrame = av_frame_alloc();
+    outputFrame->format = audio.codecCtx->sample_fmt;
+    outputFrame->sample_rate = audio.codecCtx->sample_rate;
+#if HAVE_CH_LAYOUT
+    av_channel_layout_copy(&outputFrame->ch_layout, &audio.codecCtx->ch_layout);
+#else
+    outputFrame->channel_layout = audio.codecCtx->channel_layout;
+    outputFrame->channels = audio.codecCtx->channels;
+#endif
+
+    // Use the exact frame size required by the codec
+    outputFrame->nb_samples = audio.codecCtx->frame_size;
+    av_frame_get_buffer(outputFrame, 0);
+
+    // Calculate the next PTS value based on the sample rate
+    static std::vector<int64_t> pts_counters;
+    if (pts_counters.size() <= (size_t)stream_index) {
+        pts_counters.resize(stream_index + 1, 0);
     }
 
-    int ret;
-    if ((ret = avcodec_parameters_from_context(audioStream->codecpar, audioCodecCtx)) < 0) {
-        char errmsg[256];
-        av_strerror(ret, errmsg, sizeof(errmsg));
-        std::cerr << "avcodec_parameters_from_context failed: " << err << std::endl;
-        std::exit(-1);
+    // Set PTS to the current counter value
+    outputFrame->pts = pts_counters[stream_index];
+
+    // Convert audio samples
+    int ret = swr_convert(audio.swrCtx,
+                          outputFrame->data, outputFrame->nb_samples,
+                          (const uint8_t**)inputFrame->data, inputFrame->nb_samples);
+
+    if (ret < 0) {
+        std::cerr << "Failed to convert audio samples: " << averr(ret) << std::endl;
+        av_frame_free(&inputFrame);
+        av_frame_free(&outputFrame);
+        return;
     }
+
+    // Output debug information
+    dbg << "DEBUG: Audio frame PTS for stream " << stream_index << std::endl;
+    dbg << "  - Sample rate: " << audio.codecCtx->sample_rate << std::endl;
+    dbg << "  - Audio timebase: 1/" << audio.codecCtx->sample_rate << std::endl;
+    dbg << "  - PTS: " << outputFrame->pts << std::endl;
+    dbg << "  - Time: " << (double)outputFrame->pts / audio.codecCtx->sample_rate << " seconds" << std::endl;
+
+    // Update the counter for the next frame
+    pts_counters[stream_index] += outputFrame->nb_samples;
+
+    // Send to encoder
+    AVPacket *pkt = av_packet_alloc();
+    encode(audio.codecCtx, outputFrame, pkt);
+    av_packet_free(&pkt);  // Let the encode function handle packet unref
+
+    av_frame_free(&inputFrame);
+    av_frame_free(&outputFrame);
 }
 #endif
 void FrameWriter::init_codecs()
 {
     init_video_stream();
 #ifdef HAVE_AUDIO
     if (params.enable_audio)
-        init_audio_stream();
+    {
+        // Get the number of audio sources from audioParams
```

Comment:
I am not a big fan of this, is there no way to pass the information to main.cpp and from there to framewriter?